xlunch.c

```diff
@@ -1078,33 +1082,33 @@ void run_command(char * cmd_orig)
     if (dont_quit)
     {
         pid_t pid=fork();
-        if (pid==0) // child process
-        {
+
+        switch (pid) {
+        case 0:   /* Child process */
             if (!multiple_instances) close(lock);
             printf("Forking command: %s\n",cmd);
-            int err;
-            err = execvp(array[0],array);
-            fprintf(stderr,"Error forking %s : %d\n",cmd,err);
-            exit(OKAY);
-        }
-        else if (pid<0) // error forking
-        {
-            fprintf(stderr,"Error running %s\n",cmd);
-        }
-        else // parent process
-        {
+            break;
+
+        case -1:  /* Error */
+            perror("fork");
+            /*FALLTHROUGH*/
+
+        default:  /* Parent */
             reset_prompt();
+            if (cmd != cmd_orig)
+                free(cmd);
+            return;
         }
     }
     else
     {
         cleanup();
         printf("Running command: %s\n",cmd);
-        int err;
-        err = execvp(array[0],array);
-        fprintf(stderr,"Error running %s : %d\n",cmd, err);
-        exit(OKAY);
     }
+
+    execvp(array[0],array);
+    fprintf(stderr,"Error running \"%s\": %s\n", array[0], strerror(errno));
+    exit(EXIT_FAILURE);
```

Comment:
This is again incorrect, the error code here should be `OKAY`. Xlunch has done its job and called the program, the fact that the program failed to launch isn't an error for xlunch.

Reply:
It is traditional to return an error in this situation because if you do not then the caller doesn't know that anything went wrong. This makes it difficult to invoke xlunch in a script---or to provide appropriate feedback to the user.

As an example, let's take ssh. Consider the following terminal snippet:
```
$ ssh straasha /path/to/no-where
mksh: /path/to/no-where: inaccessible or not found
$ echo $?
127
$
```
In this case, ssh did everything correctly according to its own specification. But it still returns an error because an error occurred. If you were to use ssh in a script, then this would be quite important.
A simpler example is the shell:
```
$ sh -c '/path/to/no-where'
sh: /path/to/no-where: No such file or directory
$ echo $?
127
$ 
```

Reply:
Hmm, fair enough. But at least add it as an error code in the 0x40 range, something like `EXTERNALERROR`.

Reply:
Done!